trin-core/src/portalnet/discovery.rs

```diff
@@ -139,6 +144,18 @@ impl Discovery {
         )
     }
 
+    pub async fn find_nodes(&self, enr: Enr, distances: Vec<u16>) -> Result<Vec<u8>, RequestError> {
+        let msg = FindNodes { distances };
+
+        Ok(self
+            .send_talkreq(
+                enr,
+                DISCV5_PROTOCOL.to_owned(),
+                Message::Request(Request::FindNodes(msg)).to_bytes(),
+            )
+            .await?)
+    }
```

Comment:
I'd agree, also it appears that the base `find_node` method returns a `Vec<Enr>` which is preferable to `Vec<u8>`

Reply:
To my understanding, discv5's [find_node ](https://github.com/sigp/discv5/blob/master/src/discv5.rs#L484) public method returns nodes closest to the given NodeId, which in the background is calling `nodes_by_distances` (for reference check [here](https://github.com/sigp/discv5/blob/f11a60a678c285428d47dd2b24db57633f0b3346/src/service.rs#L524) and [here](https://github.com/sigp/discv5/blob/f11a60a678c285428d47dd2b24db57633f0b3346/src/service.rs#L904)). It is not possible to pass a list of distances as parameter to this method.

On the other hand, `discv5_foundNodes` [RPC](https://playground.open-rpc.org/?schemaUrl=https://raw.githubusercontent.com/ogenev/portal-network-specs/jsonrpc-assembled-specs/jsonrpc-specs/openrpc.json&uiSchema%5BappBar%5D%5Bui:splitView%5D=false&uiSchema%5BappBar%5D%5Bui:input%5D=false&uiSchema%5BappBar%5D%5Bui:examplesDropdown%5D=false) requires from us to send a request to a target ENR with a list of a distances and is expecting to receive all nodes in the given distances from the routing table of the target node, which lead us to use the public discv5 [nodes_by_distances](https://docs.rs/discv5/0.1.0-beta.11/src/discv5/discv5.rs.html#223-245) method.

Because I think we can't use directly a discv5 `FIndNode` message, we are sending `FindNodes` request via TALKREQ and  use `discv5` as protocol identifier, this way, the main handler can distinguish between overlay and base layer discv5 `FindNodes`.

Does this makes sense? Please correct me if i'm wrong, It is possible that I'm misunderstanding something. It is kinda confusing why we use `FindNodes` for overlay, and the base discv5 message is `FindNode` (without 's).

Reply:
I think the problem we are facing here is that the public [ find_node](https://docs.rs/discv5/0.1.0-beta.11/discv5/struct.Discv5.html#method.find_node) method exposed by discv5 implementation in rust is not the actual FindNode RPC from the [specs](https://docs.rs/discv5/0.1.0-beta.11/discv5/struct.Discv5.html#method.find_node).
 The public `find_node` method is returning nodes closer to a given NodeId from the local routing table. I think the FindNode RPC message from the specs is not exposed in the public API [here](https://docs.rs/discv5/0.1.0-beta.11/discv5/struct.Discv5.html#method.find_node) (same for the PING message).